asf_levies_model/consumers.py

```diff
@@ -0,0 +1,460 @@
+                self.electricity_bill = self.electricity_subtotal_bill - (
+                    self.electricity_consumption * adjustment_parameter
+                )
+                self._adjustment_mode = adjustment_mode
+            else:
+                raise ValueError(
+                    "Please provide an adjustment mode from the following: flat adjustment, percentage_discount, unit discount"
+                )
+        elif adjustment_fuel == "gas":
+            if adjustment_mode == "flat adjustment":
+                self.gas_bill = self.gas_subtotal_bill + adjustment_parameter
+                self._adjustment_mode = adjustment_mode
+            elif adjustment_mode == "percentage discount":
+                self.gas_bill = self.gas_subtotal_bill * (
+                    (100 - adjustment_parameter) / 100
+                )
+                self._adjustment_mode = adjustment_mode
+            elif adjustment_mode == "unit discount":
+                self.gas_bill = self.gas_subtotal_bill - (
+                    self.gas_consumption * adjustment_parameter
+                )
+                self._adjustment_mode = adjustment_mode
+            else:
+                raise ValueError(
+                    "Please provide an adjustment mode from the following: flat adjustment, percentage_discount, unit discount"
+                )
+        else:
+            raise ValueError(
+                "Please specify the fuel bill to adjust (electricity or gas)."
+            )
+
+    def get_tidy_summary(self) -> pd.DataFrame:
+        """Returns a dataframe containing all public attributes and properties, except name which has a dedicated column.
+
+        Returns
+        -------
+        pd.DataFrame
+            Dataframe with columns ["Name", "Attribute", "Value"].
+        """
+        attributes = [
+            (key, value)
+            for key, value in self.__dict__.items()
+            if key != "name" and not key.startswith("_")
+        ]
+        properties = [
+            (key, getattr(self, key))
+            for key, value in self.__class__.__dict__.items()
+            if isinstance(value, property)
+        ]
+        attributes.extend(properties)
+        df = pd.DataFrame(attributes, columns=["Attribute", "Value"])
+        df["Name"] = self.name
+        return df[["Name", "Attribute", "Value"]]
+
+    @classmethod
+    def consumer_from_dataframe(
+        cls,
+        df: pd.DataFrame,
+        row: int,
+        name_col: str,
+        archetype_col: str,
+        net_annual_income_col: str,
+        main_heating_fuel_col: str,
+        gas_consumption_col: str,
+        electricity_consumption_col: str,
+        gas_tariff: object,
+        electricity_tariff: object,
+        net_income_decile_col: str = None,
+        unmetered_fuel_spend_col: str = None,
+        scheme_eligible_col: str = None,
+        size_col: str = None,
+        unit_converter: float = 1,
+    ) -> "Consumer":
```

Comment:
Things which default to None but have a different datatype can be given the `Optional` type hint.

If a number has a float type and has a default, the default should be a float, not an int (i.e. 1.0 not 1)